src/mesh/amr_loadbalance.cpp

```diff
@@ -70,73 +72,94 @@ void Mesh::LoadBalancingAndAdaptiveMeshRefinement(ParameterInput *pin) {
-// \!fn void Mesh::CalculateLoadBalance(double *clist, int *rlist, int *slist,
-//                                      int *nlist, int nb)
-// \brief Calculate distribution of MeshBlocks based on the cost list
-
-void Mesh::CalculateLoadBalance(double *clist, int *rlist, int *slist, int *nlist,
-                                int nb) {
-  std::stringstream msg;
-  double real_max = std::numeric_limits<double>::max();
-  double totalcost = 0, maxcost = 0.0, mincost = (real_max);
-
-  for (int i = 0; i < nb; i++) {
-    totalcost += clist[i];
-    mincost = std::min(mincost, clist[i]);
-    maxcost = std::max(maxcost, clist[i]);
-  }
-
-  int j = (Globals::nranks)-1;
-  double targetcost = totalcost / Globals::nranks;
-  double mycost = 0.0;
+// Private routines
+namespace {
+/**
+ * @brief This routine assigns blocks to ranks by attempting to place index-contiguous
+ * blocks of equal total cost on each rank.
+ *
+ * @param costlist (Input) A map of global block ID to a relative weight.
+ * @param ranklist (Output) A map of global block ID to ranks.
+ */
+void AssignBlocks(std::vector<double> const &costlist, std::vector<int> &ranklist) {
+  ranklist.resize(costlist.size());
+
+  double const total_cost = std::accumulate(costlist.begin(), costlist.end(), 0.0);
+
+  int rank = (Globals::nranks)-1;
+  double target_cost = total_cost / Globals::nranks;
+  double my_cost = 0.0;
+  double remaining_cost = total_cost;
   // create rank list from the end: the master MPI rank should have less load
-  for (int i = nb - 1; i >= 0; i--) {
-    if (targetcost == 0.0) {
+  for (int block_id = costlist.size() - 1; block_id >= 0; block_id--) {
+    if (target_cost == 0.0) {
+      std::stringstream msg;
       msg << "### FATAL ERROR in CalculateLoadBalance" << std::endl
           << "There is at least one process which has no MeshBlock" << std::endl
           << "Decrease the number of processes or use smaller MeshBlocks." << std::endl;
       PARTHENON_FAIL(msg);
     }
-    mycost += clist[i];
-    rlist[i] = j;
-    if (mycost >= targetcost && j > 0) {
-      j--;
-      totalcost -= mycost;
-      mycost = 0.0;
-      targetcost = totalcost / (j + 1);
-    }
-  }
-  slist[0] = 0;
-  j = 0;
-  for (int i = 1; i < nb; i++) { // make the list of nbstart and nblocks
-    if (rlist[i] != rlist[i - 1]) {
-      nlist[j] = i - slist[j];
-      slist[++j] = i;
+    my_cost += costlist[block_id];
+    ranklist[block_id] = rank;
+    if (my_cost >= target_cost && rank > 0) {
+      rank--;
+      remaining_cost -= my_cost;
+      my_cost = 0.0;
+      target_cost = remaining_cost / (rank + 1);
     }
   }
-  nlist[j] = nb - slist[j];
+}
 
-  if (Globals::my_rank == 0) {
-    for (int i = 0; i < Globals::nranks; i++) {
-      double rcost = 0.0;
-      for (int n = slist[i]; n < slist[i] + nlist[i]; n++)
-        rcost += clist[n];
+void UpdateBlockList(std::vector<int> const &ranklist, std::vector<int> &nslist,
+                     std::vector<int> &nblist) {
```

Comment:
### Optional

Is this function completely over writing nslist and nblist? I'm not sure, but if it is I would think it would be preferable to simply return ns and nb list in the return value. This might be more stylistic than necessary though.

Reply:
Let's pass on this for now. I really want to get this thing merged in, and it's been sitting for a long time.